test/index.spec.js

```diff
@@ -0,0 +1,212 @@
+const timeoutError = { code: 'ETIMEDOUT', name: 'TimeoutError' };
+const cancelledError = { code: 'ECANCELLED', name: 'CancelledError' };
+const asyncActionsCreator = (actions, timeoutMs = 10) => (dispatch) => {
+    const timeout = setTimeout(() => {
+        clearTimeout(timeout);
+        actions.forEach((action) => dispatch(action));
+    }, timeoutMs);
+};
+const actionsCreator = (actions) => (dispatch) =>
+    actions.forEach((action) => dispatch(action));
+const createMockStore = (middlewares = []) => configureStore(middlewares)(initialState);
+
+afterEach(() => {
+    jest.useRealTimers();
+    jest.clearAllMocks();
+});
+
+it('should resolve the promise when a single action is dispatched', async () => {
+    const mockStore = createMockStore();
+
+    mockStore.dispatch(action1);
+    await waitForActions(mockStore, action1);
+
+    expect(mockStore.getActions()).toEqual([action1]);
+});
+
+it('should resolve the promise when action creator is dispatched', async () => {
+    const mockStore = createMockStore([thunkMiddleware]);
+    const actions = [action1, action2, action3];
+
+    mockStore.dispatch(actionsCreator(actions));
+    await waitForActions(mockStore, actions);
+
+    expect(mockStore.getActions()).toEqual(actions);
+});
+
+it('should resolve the promise when async action creator is dispatched', async () => {
+    const mockStore = createMockStore([thunkMiddleware]);
+    const actions = [action1, action2, action3];
+
+    mockStore.dispatch(asyncActionsCreator(actions));
+    await waitForActions(mockStore, actions);
+
+    expect(mockStore.getActions()).toEqual(actions);
+});
+
+it('should resolve the promise when expected actions match a subset of property values of dispatched actions', async () => {
+    const mockStore = createMockStore([thunkMiddleware]);
+    const actions = [action1, action2, action3, action4];
+
+    mockStore.dispatch(actionsCreator(actions));
+    await waitForActions(mockStore, [
+        { type: 'ACTION-1', payload: { id: 1 } },
+        { type: 'ACTION-2', payload: { name: 'ACTION TWO' } },
+        { type: 'ACTION-3' },
+        { type: 'ACTION-4', payload: { id: 4, name: 'ACTION FOUR' } },
+    ]);
+});
+
+it('should resolve the promise when no actions are expected', async () => {
+    const mockStore = createMockStore();
+    const actions = [];
+
+    await waitForActions(mockStore, actions);
+
+    expect(mockStore.getActions()).toEqual(actions);
+});
+
+it('should reject the promise when expected actions are not received and timeout expires', async () => {
+    const mockStore = createMockStore();
+    const actions = [action1, action2, action3];
+
+    mockStore.dispatch(action1);
+    mockStore.dispatch(action2);
+
+    let error;
+
+    try {
+        await waitForActions(mockStore, actions);
+    } catch (e) {
+        error = e;
+    }
+
+    expect(error).toBeDefined();
+    expect({ code: error.code, name: error.name }).toEqual(timeoutError);
```

Comment:
When testing async code, you can use `resolve/rejects`:

```js
const mockStore = createMockStore();
const actions = [action1, action2, action3];

mockStore.dispatch(action1);
mockStore.dispatch(action2);

await expect(waitForActions(mockStore, actions)).rejects.toEqual(timeoutError);
```

http://facebook.github.io/jest/docs/en/asynchronous.html#async-await

Reply:
I wanted to assert that the error thrown has properties `code` and `name` with the expected values. The code above does work in this case. The matcher `.toMatchObject()` has to be used instead. However, `.toMatchObject()` does not match `Errors` with properties anymore since Jest v21.0+. That is why I opted for a more verbose approach to workaround the problem. Meanwhile, I changed the dependency of Jest to v20.

There is a [issue](https://github.com/facebook/jest/issues/4532) currently opened in Jest's repo reporting this change of behavior.